bugsnag-plugin-android-ndk/src/main/jni/ctjournal.c

```diff
@@ -0,0 +1,587 @@
+ * `{"some.path":` part. Android 4.0 (API 14) default stack size is 16k:
+ * https://android.googlesource.com/platform/dalvik/+/android-cts-4.0_r1/vm/Thread.h#79
+ * We need to me mindful of stack space, so pick a value that won't be blown out
+ * by the deepest document path we use, and also won't put much pressure on the
+ * stack.
+ */
+#define JOURNAL_BEGIN_BUFFER_SIZE 200
+
+/* IMPORTANT: Keep consistent with:
+ * - bugsnag_breadcrumb_type in
+ *       bugsnag-plugin-android-ndk/src/main/assets/include/event.h
+ * - BreadcrumbType in
+ *       bugsnag-android-core/src/main/java/com/bugsnag/android/BreadcrumbType.kt
+ */
+static const char *g_breadcrumb_types[] = {
+    "manual",     // BSG_CRUMB_MANUAL
+    "error",      // BSG_CRUMB_ERROR
+    "log",        // BSG_CRUMB_LOG
+    "navigation", // BSG_CRUMB_NAVIGATION
+    "process",    // BSG_CRUMB_PROCESS
+    "requsest",   // BSG_CRUMB_REQUEST
+    "state",      // BSG_CRUMB_STATE
+    "user",       // BSG_CRUMB_USER
+};
+static const int g_breadcrumb_types_count =
+    sizeof(g_breadcrumb_types) / sizeof(*g_breadcrumb_types);
+
+// Single journal file descriptor (there can only be one crash-time journal)
+static int g_fd = 0;
+
+// Path to journal (only used for error reporting)
+static char g_journal_path[PATH_MAX];
+
+/**
+ * Convenience macro to stop the current function and return false if a call
+ * returns false.
+ */
+#define RETURN_ON_FALSE(...)                                                   \
+  do {                                                                         \
+    if (!__VA_ARGS__) {                                                        \
+      return false;                                                            \
+    }                                                                          \
+  } while (0)
+
+static inline bool write_bytes(const char *bytes, size_t length) {
+  if (write(g_fd, bytes, length) < 0) {
+    BUGSNAG_LOG("Error writing to journal at %s: %s", g_journal_path,
+                strerror(errno));
+    return false;
+  }
+  return true;
+}
+
+static inline bool write_string(const char *string) {
+  return write_bytes(string, strlen(string));
+}
+
+/**
+ * This is implemented as a macro to take advantage of compile-time size
+ * calculations when passing in string literals.
+ */
+#define write_literal(LITERAL_STRING)                                          \
+  write_bytes((LITERAL_STRING), sizeof(LITERAL_STRING) - 1)
+
+static inline bool write_boolean(bool value) {
+  if (value) {
+    return write_literal("true");
+  } else {
+    return write_literal("false");
+  }
+}
+
+static inline bool write_null() { return write_literal("null"); }
+
+// The following, including uint64_to_string(), int64_to_string(),
+// positive_double_to_string(), and double_to_string() were copied directly
+// from:
+// https://github.com/bugsnag/bugsnag-cocoa/blob/master/Bugsnag/KSCrash/Source/KSCrash/Recording/Tools/BSG_KSJSONCodec.c#L90
```

Comment:
This big chunk comes directly from bugsnag-cocoa.

Reply:
It might be worth extracting the functions to a separate file to make the separation obvious, and reduce the source file length